### PR TITLE
[2/2] refactor: decoupled self distillation trainers; cleanup

### PR DESCRIPTION
**Pull request by LeonEricsson**
# What does this PR do?

Follow up #5862. 
A SDPO loss change and a bunch of non-behavior changing refactors

## Changes

1. **SDPO loss → convex combination**. Replaces the additive `policy + λ·distillation` with the paper's `(1 - w)·policy + w·distillation` (Section 4.5). `sdpo_policy_loss_mode` is removed; `distillation_weight` is now the convex weight `w ∈ [0, 1]` (`1.0` = pure distillation = prior default, `0.0` = pure policy gradient)
2. **Config docstrings** completed for `SDFTConfig`/`SDPOConfig`, removed unused diagnostics from SDFT.
3. **Metric logging** few helpers that move logging out of core functions to make things more readable
4. **Method reordering** moved around some methods in SDFT/SDPO.
5. **FSDP fix**: wrap transformers generation in `summon_full_params`.
6. **loss utils**: docstrings improved, cleanup dead code

#### notes
item 1 is behavioral change; the rest is docs/structure/fixes.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> SDPO training loss semantics change for hybrid runs (additive weight vs convex blend), and configs using removed `sdpo_policy_loss_mode` must migrate to `distillation_weight`.
> 
> **Overview**
> **SDPO** replaces `sdpo_policy_loss_mode` (`distillation_only` / `hybrid`) with a single **`distillation_weight`** in `[0, 1]` that convex-combines objectives: `(1 - w) * policy_loss + w * distillation_loss` (`1.0` = pure distillation, `0.0` = GRPO-style policy only). Liger now requires `distillation_weight=1.0`. Config validates `distillation_weight`; docs and CLI examples use `--distillation_weight` instead of `--sdpo_policy_loss_mode`.
> 
> **SDFT/SDPO trainers** rename distillation masking from `response_mask` to **`loss_mask`** in `DistillationLogits`; SDFT drops per-sample `self_distillation_mask` from batches and applies `num_loss_tokens_to_skip` directly on `completion_mask`. Shared **`loss_utils`** dedupe token log-probs via `selective_log_softmax`, rename tail handling to `add_tail_bucket`, and document top-k tail vs renormalize behavior.
> 
> **SDFTConfig** gains full parameter docstrings (aligned with SDPO), default **`disable_dropout=True`**, and removes unused diagnostics fields. **Generation** under FSDP wraps `model.generate` in **`FSDP.summon_full_params`**. Completion metrics move to **`_record_completion_metrics`**; SDPO **`compute_loss`** is simplified (no separate hybrid helpers). Tests updated for the new masks and `distillation_weight`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f00e139362626fbbfcd874c0b70ca13d25e485a9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/sdft/sdft_trainer.py [hunk 1]
```diff
@@ -122,36 +137,40 @@ def build(
...
     ) -> dict[str, torch.Tensor]:
-        student_batch = self.prompt_tokenizer.tokenize_prompts(prompts)
         teacher_prompts = [
             self._compose_teacher_prompt(prompt, privileged_context)
             for prompt, privileged_context in zip(prompts, privileged_contexts, strict=True)
         ]
-        teacher_batch = self.prompt_tokenizer.tokenize_prompts(teacher_prompts)
-        teacher_input_ids = torch.cat([teacher_batch.prompt_ids, completion_ids], dim=1)
-        teacher_attention_mask = torch.cat([teacher_batch.prompt_mask, completion_mask], dim=1)
+        teacher_prompt_ids_list = self.trainer._tokenize_prompts(teacher_prompts)
```
[the remark this answers is not in the document]
Comment (LeonEricsson) [reply trl/experimental/sdft/sdft_trainer.py:161]: This is the same behavior as on `main`, not introduced or changed by this PR.

SDFT's privileged context is dataset-provided and typically bounded, so the original solution reuses `max_prompt_length` for both student and teacher prompts. SDPO's teacher prompt is more dynamic; it can embed a full completion among other things, which generally makes the re-prompted teacher input considerably longer than the student prompt. As a result, SDPO's teacher is left-truncated via `max_reprompt_len`.

Both are left-truncated, just using different params.

**Comment by kashif** [comment]
Went through this and ran it on 2 GPUs. The convex loss matches the paper (section 4.5) and what we discussed earlier, fsdp2 trains fine at both `distillation_weight=0.5` and `1.0`, and the experimental tests pass (34). One small thing I noticed: `distillation_weight` is documented as `[0, 1]` but nothing enforced it, so a value like 1.5 silently makes the policy term negative. Pushed a tiny guard in `__post_init__` for it. Everything else looks good to me, and the fsdp summon matches what grpo/rloo already do. LGTM 👍

**Review by kashif — APPROVED** [review]
